(MODULES-6800) Add Windows service management

Iristyle opened the pull request:
- Basic ability to start / stop / restart and query status on
   a service given a name

 - Follows the same model as existing task where the state machine
   flows like:

   - status
     status: started
   - start
     status: in_sync
   - restart
     status: restarted
   - status
     status: started
   - stop
     status: stopped

   - status
     status: stopped
   - stop
     status: in_sync
   - restart
     status: restarted

   - status
     status: started

   NOTE: Windows `Running` status is converted to `started`

 - For states Puppet doesn't currently model, but that Windows will
   return, return those values:

   ContinuePending
   Paused
   PausePending
   StartPending
   StopPending

Iristyle commented:
Tested manually against 2012R2:

#### action=status

```
bundle exec bolt task run service::windows action=status name=lmhosts --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on d5kfl7ziuu5fcio.delivery.puppetlabs.net...
Finished on d5kfl7ziuu5fcio.delivery.puppetlabs.net:
  {
    "Name": "lmhosts",
    "Status": "Running"
  }
Successful on 1 node: winrm://d5kfl7ziuu5fcio.delivery.puppetlabs.net
Ran on 1 node in 1.98 seconds
```

#### action=restart

```
bundle exec bolt task run service::windows action=restart name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on d5kfl7ziuu5fcio.delivery.puppetlabs.net...
Finished on d5kfl7ziuu5fcio.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Running"
  }
Successful on 1 node: winrm://d5kfl7ziuu5fcio.delivery.puppetlabs.net
Ran on 1 node in 2.36 seconds
```

#### action=stop

```
bundle exec bolt task run service::windows action=stop name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on d5kfl7ziuu5fcio.delivery.puppetlabs.net...
Finished on d5kfl7ziuu5fcio.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Stopped"
  }
Successful on 1 node: winrm://d5kfl7ziuu5fcio.delivery.puppetlabs.net
Ran on 1 node in 2.04 seconds
```

followed by a status for good measure

```
bundle exec bolt task run service::windows action=status name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on d5kfl7ziuu5fcio.delivery.puppetlabs.net...
Finished on d5kfl7ziuu5fcio.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Stopped"
  }
Successful on 1 node: winrm://d5kfl7ziuu5fcio.delivery.puppetlabs.net
Ran on 1 node in 1.98 seconds
```

#### action=start

```
bundle exec bolt task run service::windows action=start name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on d5kfl7ziuu5fcio.delivery.puppetlabs.net...
Finished on d5kfl7ziuu5fcio.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Running"
  }
Successful on 1 node: winrm://d5kfl7ziuu5fcio.delivery.puppetlabs.net
Ran on 1 node in 2.38 seconds
```

Iristyle commented:
A few open questions:

* Should we just reuse the existing `init.json` and add an `init.ps1`? This would make sense to me for a Windows implementation where the agent is not necessary at all (which is different from the more complex scenario for #23), so that we still maintain a "generic" service control interface task.
* Not sure yet how this is intended to be tested in Jenkins... and AppVeyor isn't setup here.
* Any way to make it clear in the metadata that `Provider` is not necessary on Windows?
* Enable / Disable don't make sense in Windows context so aren't implemented - punt and ticket for later?

Iristyle commented:
In secondary commit implemented setting enabled / disabled variants. This is a bit shoe-horned in to the `action` parameter, even though it could be modeled as a "starttype" parameter.  This was done for now to simplify parameters / matrix of commands, but could be changed.

#### action=enable_manual

```
bundle exec bolt task run service::windows action=enable_manual name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on iqiwp2g7zzvlvb2.delivery.puppetlabs.net...
Finished on iqiwp2g7zzvlvb2.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Running",
    "StartType": "Manual"
  }
Successful on 1 node: winrm://iqiwp2g7zzvlvb2.delivery.puppetlabs.net
Ran on 1 node in 2.13 seconds
```

#### action=enable_automatic

```
bundle exec bolt task run service::windows action=enable_automatic name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on iqiwp2g7zzvlvb2.delivery.puppetlabs.net...
Finished on iqiwp2g7zzvlvb2.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Running",
    "StartType": "Automatic"
  }
Successful on 1 node: winrm://iqiwp2g7zzvlvb2.delivery.puppetlabs.net
Ran on 1 node in 2.06 seconds
```

#### action=disable

```
bundle exec bolt task run service::windows action=disable name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on iqiwp2g7zzvlvb2.delivery.puppetlabs.net...
Finished on iqiwp2g7zzvlvb2.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Running",
    "StartType": "Disabled"
  }
Successful on 1 node: winrm://iqiwp2g7zzvlvb2.delivery.puppetlabs.net
Ran on 1 node in 2.19 seconds
```

glennsarti commented on tasks/windows.json:
```diff
@@ -0,0 +1,14 @@
+      "type": "Enum[start, stop, restart, enable_automatic, enable_manual, disable, status]"
+    },
+    "name": {
+      "description": "The name of the service to operate on.",
```
May be worth adding a  note here that the service name is Short version of the name, not the displayname.

Iristyle commented:
Latest commit allows for combining `starttype` and `action` in one task execution, enabling some more interesting combinations like:

#### action=start / starttype=automatic

```
bundle exec bolt task run service::windows action=start starttype=automatic name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on iqiwp2g7zzvlvb2.delivery.puppetlabs.net...
Finished on iqiwp2g7zzvlvb2.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Running",
    "StartType": "Automatic"
  }
Successful on 1 node: winrm://iqiwp2g7zzvlvb2.delivery.puppetlabs.net
Ran on 1 node in 2.17 seconds
```

#### action=start / starttype=disabled

```
bundle exec bolt task run service::windows action=start starttype=disabled name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on iqiwp2g7zzvlvb2.delivery.puppetlabs.net...
Finished on iqiwp2g7zzvlvb2.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Running",
    "StartType": "Disabled"
  }
Successful on 1 node: winrm://iqiwp2g7zzvlvb2.delivery.puppetlabs.net
Ran on 1 node in 2.16 seconds
```

#### action=status / starttype=disabled

```
bundle exec bolt task run service::windows action=status starttype=disabled name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on iqiwp2g7zzvlvb2.delivery.puppetlabs.net...
Finished on iqiwp2g7zzvlvb2.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Running",
    "StartType": "Disabled"
  }
Successful on 1 node: winrm://iqiwp2g7zzvlvb2.delivery.puppetlabs.net
Ran on 1 node in 2.21 seconds
```

#### action=stop / starttype=automatic

```
bundle exec bolt task run service::windows action=stop starttype=automatic name=VMTools --nodes $BOLT_HOST --user $BOLT_USER --password $BOLT_PASSWORD --modulepath ~/source/bolt-tasks --no-ssl
Started on iqiwp2g7zzvlvb2.delivery.puppetlabs.net...
Finished on iqiwp2g7zzvlvb2.delivery.puppetlabs.net:
  {
    "Name": "VMTools",
    "Status": "Stopped",
    "StartType": "Automatic"
  }
Successful on 1 node: winrm://iqiwp2g7zzvlvb2.delivery.puppetlabs.net
Ran on 1 node in 2.16 seconds
```

glennsarti commented:
Not a blocker, but the more complex startup types will not be shown i.e.  `Automatic (Delayed)` and `Manual (Triggered)` will appear as `Automatic` and `Manual`.  This is a PowerShell limitation not bolt

glennsarti commented:
@Iristyle There is currently not a testing story for tasks.  Is suspect you'd need to write pester tests instead

glennsarti commented on tasks/windows.json:
```diff
@@ -0,0 +1,18 @@
+{
+  "description": "Manage the state of Windows services (without a puppet agent)",
```
Is the "Without a puppet agent" needed?  Isn't bolt implicitly sans-puppet

Iristyle replied:
Nope... same tasks can be run through PE tasks support via pxp-agent OR via Bolt. The other existing `init.sh` task in this module uses the Puppet agent.

Also, I did crib that bit from the TP PR at #23

glennsarti commented on tasks/windows.json:
```diff
@@ -0,0 +1,18 @@
+      "description": "The operation (start, stop, restart, status) to perform on the service",
+      "type": "Enum[start, stop, restart, status]"
+    },
+    "starttype": {
```
nit: Should this parameter name indicate an action will place.  e.g. `set-starttype`

Iristyle replied:
Maybe - will see if there's a naming convention in place elsewhere.

jpogran replied:
I think 'action' is implied, in that this is a task which is implied it changes state.

Iristyle replied:
So I had started with jamming startup type into the `action` to experiment with the task interaction, but after playing with that a bit, thought that was a bad fit, and makes some operations that could be one step into two instead. The third commit moved out `starttype`, which is more in sync with how it's modeled on Windows / PowerShell cmdlets / etc... but we also end up with one weird scenario as a direct result, which is why I was leaning away from doing this from the beginning.

Setting an `action` of `status` and then also setting `starttype` to any of the values will modify the startup type, which seems a bit counter-intuitive as a consumer of the task.  BUT, it still might be OK / something that can be documented around / something that could be outright rejected in the task. Could also outright remove `status`, but I think that's something that Discovery is interested in querying.

Interested in hearing what you think @jpogran / @glennsarti?

jpogran replied:
I think it's a separate cmdlet for a reason. Setting start type is a different action that setting service state, and one can operate without the other.

I understand wanting to hit the box the least amount of times you can, but IMHO there isn't sufficient error handling here to prevent a user from setting a state that conflicts with a start type. An error setting state will kill the execution, missing a valid setting for start type. I realize that tasks aren't supposed to be idempotent per se, but by conflating the two an error in one leaves the node in an unknown state. You could spend alot of code handling different situations and returning information based on different outcomes, or you could make two tasks.

Iristyle commented:
@glennsarti I had punted on the `delay auto` and `manual trigger` services since they are special variations of `automatic` and `manual` configured with the [ChangeServiceConfig2 API](https://msdn.microsoft.com/en-us/library/windows/desktop/ms681988\(v=vs.85\).aspx) (not their own startup types - see [the 5 start types in Win32 API](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682450\(v=vs.85\).aspx)).

I think the ChangeServiceConfig2 API stashes settings in the registry, in which case we could plausibly find a way to support them, even without calling the Windows API, in order to avoid using PInvoke.

* [SERVICE_DELAYED_AUTO_START_INFO](https://msdn.microsoft.com/en-us/library/windows/desktop/ms685155\(v=vs.85\).aspx) only takes a bool value for on / off, so that should be pretty straightforward.
* [SERVICE_TRIGGER_INFO](https://msdn.microsoft.com/en-us/library/windows/desktop/dd405514\(v=vs.85\).aspx) is a bit more complicated, but I think the type system for metadata could probably map the [SERVICE_TRIGGER](https://msdn.microsoft.com/en-us/library/windows/desktop/dd405512\(v=vs.85\).aspx) structures. 

We'd also need to grab the information when emitting the JSON to synthesize those values, and this makes the validation logic a bit more complex.

It can be done for sure, but wonder if it's necessary (yet)?

jpogran commented on tasks/windows.ps1:
```diff
@@ -0,0 +1,46 @@
+
+$service = Get-Service -Name $Name
+
+switch ($StartType.ToLower())
```
powershell switch statement is case-insensitive by default, so tolower (while stating intent clearly) is redundant

jpogran replied:
if tolower is kept, consider tolowerinvariant as that won't change user input

Iristyle replied:
Thanks - forgot about that! Also meant to put a TODO there to double check how the Puppet type system metadata works... The `[Enum]` may already be case sensitive, in which case some of the validation here is moot.

jpogran commented on tasks/windows.json:
```diff
@@ -0,0 +1,18 @@
+      "type": "Optional[Enum[automatic, manual, disabled]]"
+    },
+    "name": {
+      "description": "The name of the service to operate on.",
```
Related to @glennsarti comments earlier, I would specify here that this is the service `name` and call out that it is not the `displayname`, to avoid user confusion.

Iristyle replied:
Will update metadata docs

jpogran commented on tasks/windows.json:
```diff
@@ -0,0 +1,18 @@
+  "description": "Manage the state of Windows services (without a puppet agent)",
+  "input_method": "powershell",
+  "parameters": {
+    "action": {
```
I'm torn on the name `action`. My first expectation from prior experience is that this should be `state`, as everything I've ever used goes that way. This does not account for 'status', but I am used to separate commands for getting status and setting status. I am unaware of what the current conventions for using tasks, so ignore if the other tasks use this convention

Iristyle replied:
Yeah, not sure how much thought was given to `action`, but that was already in use for the existing task, so I carried it forward - https://github.com/puppetlabs/puppetlabs-service/blob/master/tasks/init.json#L5

Just checked on package tasks and they follow the same structure - https://github.com/puppetlabs/puppetlabs-package/blob/master/tasks/init.json#L5-L8

jpogran commented on tasks/windows.ps1:
```diff
@@ -0,0 +1,46 @@
+  $Name,
+
+  [Parameter(Mandatory = $true)]
+  [ValidateSet('start', 'stop', 'restart', 'status')]
```
You could avoid the tolower calls later on by using IgnoreCase: `[ValidateSet('start', 'stop', 'restart', 'status', IgnoreCase = $false)]`, which would require the case you specify. However this throws a `ParameterBindingValidationException` which kills execution before getting into the script if you use any variation of case other than what's in the set. This may be more user hostile than you intend.

Iristyle replied:
Just double checked - Puppet language definition of `Enum` is case-sensitive already per https://puppet.com/docs/puppet/5.3/lang_data_abstract.html#enum so setting `IgnoreCase = $false` would match how Bolt / PE tasks will be called by the system.

That said, that means it isn't super necessary one way or another either. (Even the `[ValidateSet]` is not protecting anything - just adding more code to keep in sync with `metadata.json`. I'll vet a couple scenarios locally where things are mismatched.)

As for the UX part, I believe the web interface uses the `Enum` values to fill a drop-down list, so that a user selects the value from a list rather than does raw data entry... at the CLI `bolt task show service::windows` should give similar info to callers, and validation should happen prior to calling the PS.

Iristyle commented on tasks/windows.ps1:
```diff
@@ -0,0 +1,82 @@
+  Write-Host @"
+  {
+    "status"  : "failure",
+    "error"  : {
```
The original task uses `_error`, but Bolt is unable to properly parse the return value with the `_` present in PowerShell

For instance, given  `Write-Host '{ "error" : true }'` the response is:

```
Started on 127.0.0.1...
Finished on 127.0.0.1:
  {
    "error": true
  }
Successful on 1 node: winrm://127.0.0.1:55985
Ran on 1 node in 4.77 seconds
```

However, changing to `Write-Host '{ "_error" : true }'`, Bolt fails like

```
Started on 127.0.0.1...
Failed on 127.0.0.1:
Failed on 1 node: winrm://127.0.0.1:55985
Ran on 1 node in 4.71 seconds
```

/cc @MikaelSmith  @nicklewis

Iristyle replied:
@MikaelSmith says what I'm seeing is by design, though it was a little surprising to me.

I have changed it back to `_error`.

Iristyle commented on tasks/windows.ps1:
```diff
@@ -0,0 +1,82 @@
+  Write-Host @"
+  {
+    "Name"        : "$($service.Name)",
+    "DisplayName" : "$($service.DisplayName)",
```
Beware of service names with embedded `"`

glennsarti replied:
If only we had use of ConvertTo-JSON :-(

Iristyle replied:
Yup, I was already thinking that before you wrote it ;0

As a follow-up, I've filed https://tickets.puppetlabs.com/browse/BOLT-406 because I believe this functionality should be available everywhere to task authors.

Iristyle commented on tasks/windows.ps1:
```diff
@@ -0,0 +1,82 @@
+
+function Invoke-ServiceAction($Service, $Action)
+{
+  $inSyncStatus = 'in_sync'
```
Rather than returning `in_sync`, I think it would be more useful to return the initial service state and the end service state after performing an action. I followed the model of the existing Puppet based task, but returning `restarted` is a bit nebulous, for instance - given a service start could fail, but this may not  be accurately represented in the return `status` (note that a failure to start should be an `Exception` which is currently handled and returned in an `_error` object, but there may be a situation where a start results in `StartPending`).

/cc @ccaum / @kenazk / @tylerpace

ccaum replied:
+1 I'll raise this issue in the main Puppet based one too

Iristyle replied:
@ccaum should we get a ticket filed for this / a PR spiked?

Iristyle replied:
I filed https://tickets.puppetlabs.com/browse/MODULES-6931 to track the normalization. If we decide *not* to do that, we can close it with a note.

Iristyle replied:
I tossed up a quick spike at https://github.com/puppetlabs/puppetlabs-service/pull/29

Iristyle commented on tasks/windows.ps1:
```diff
@@ -0,0 +1,82 @@
+{
+  Write-Host @"
+  {
+    "status"  : "failure",
```
Seems like failure payload should still contain the service `name` and the attempted `action` to be fully self-describing.

glennsarti commented on tasks/windows.ps1:
```diff
@@ -0,0 +1,82 @@
+  {
+    'start'
+    {
+      if ($Service.Status -eq 'Running') { $status = $InSyncStatus }
```
What about if it's in StartPending?

Iristyle replied:
Given the current specification that does not match `Running`

This is one of the reasons I think that we should rethink the responses, removing the `in_sync` status. Better to give the initial service state and the service state after the `action` is performed and let the caller exercise it's own logic around how to surface that to users.

glennsarti commented on tasks/windows.ps1:
```diff
@@ -0,0 +1,82 @@
+    }
+    'stop'
+    {
+      if ($Service.Status -eq 'Stopped') { $status = $InSyncStatus }
```
What about if it's in StopPending?

Iristyle replied:
Same comment as https://github.com/puppetlabs/puppetlabs-service/pull/27#discussion_r177525854

Iristyle commented:
Made some minor tweaks and I think this is ready to go -

* Made all JSON keys camelCase
* Added `"action"` to the success payload so that the message is entirely self-describing
* Similarly, added the incoming `"name"` and `"action"` to the failure payload for the same reasons
* Improved the failure message so that it includes the text `"Unable to perform '$Action' on '$Name':"` prior to actual Exception message

Have verified locally against Nano server.

Iristyle commented on tasks/windows.ps1:
```diff
@@ -0,0 +1,85 @@
+    "_error"  : {
+      "msg" : "Unable to perform '$Action' on '$Name': $($_.Exception.Message)",
+      "kind": "powershell_error",
+      "details" : {}
```
I considered putting the `$_.Exception.Source` / `$_.Exception.StackTrace` in details, but ... didn't want to deal with the embedded newlines in JSON. Also didn't think anything would be available to consume that info.
